### PR TITLE
[XLA] HLO Executor based backend example

### PR DESCRIPTION
**Pull request by DavidNorman**
Adding an example XLA device that uses the hlo evaluator to perform the execution.

### CODE REVIEW

**Comment by DavidNorman** [comment]
@kayzhu @eliben 

Decided to make a new squashed version instead of the merge that included a whole lot of other stuff.

It passes some of the XLA unit tests (you would need pull request https://github.com/tensorflow/tensorflow/pull/9759 to be able to run the unit tests).

It crashes on a couple, and it hangs on the variable test, not sure why, breaking into it shows the evaluator executing binary ops.  but mostly it just complains errors out with unsupported ops.

nevertheless - this pull request could be useful to you.

**Comment by eliben** [comment]
Is "plugin/example" still the best name for this backend? Maybe call it something like "evaluator"?

**Comment by DavidNorman** [comment]
good point

**Comment by kayzhu** [comment]
Very nice! I am still reading (sorry for the delay) so more to come :)

One immediate observation I have is I don't quite understand the meaning of "plugin" in this context. It seems me a more fitting place would be a executor/ backend similar to CPU/GPU, under service/ directory, but I will defer the final say to eliben@.

**Comment by DavidNorman** [comment]
It was originally a thing I created when I was trying to bootstrap my backend.  Originally we were hoping (as suggested by vrv) to make it into an actual loadable plugin.  However protobuf isn't really up to the job of working in that way, so I abandoned that earlier.  Nevertheless, the files remained in a directory called plugin.

I quite like the isolation of the driver from the rest of XLA.

But it is as you wish.  Feel free to take this and run with it.  I've got plenty to do on my own company's backend really.  I don't suppose that they would be too happy for me to spend time on other things.

File: tensorflow/compiler/plugin/executor/device.cc
Comment (kayzhu) [inline tensorflow/compiler/plugin/executor/device.cc]: Missing DT_INT64 and DT_DOUBLE

**Comment by kayzhu** [comment]
@tensorflow-jenkins test this please

**Comment by kayzhu** [comment]
Thanks again for your contribution! I understand that this is a rather large, but otherwise self-contained change, so I'd like to reduce the burden for you as much as possible.

In that light, what do you think of we proceed as follows:
1. fixing up the presubmit failures.
2. fix potential major issues, if any.
3. merge, and we can then fix up minor issues, if any
?

It's probably worth pointing out, we *may* later change the location/structure of this plugin, within reasonable bound. If your internal code are dependents of them, they will eventually need to be modified as well to some degree.

**Comment by DavidNorman** [comment]
sure.   my backend doe not rely on this at all, although it is of a very similar structure in terms of file layout in the repo.

**Comment by DavidNorman** [comment]
I updated the API to match the current repo changes, and added your extra data types.

**Review by kayzhu — COMMENTED** [review]
A few quick comments for first pass -- hopefully not much more in the next round. Thanks for your time in advance in addressing them!

**Comment by DavidNorman** [comment]
Incidentally, when you run the CPU tests (or any test?) at the moment you get errors like these:

```
2017-05-26 09:25:47.991881: E tensorflow/compiler/xla/service/hlo_evaluator.cc:582] TryEvaluate failed:Unimplemented: unhandled HLO ops for HloEvaluator: pad.
```
I think that this is something to do with having it included in one of the optimizers?

I don't think that the message should be listed as an error if it is going to appear regularly.

**Comment by kayzhu** [comment]
Yes you are right, the error message that change has been reverted, and will be a VLOG going forward once it is re-committed.

**Comment by kayzhu** [comment]
@tensorflow-jenkins test this please

**Comment by DavidNorman** [comment]
the error is in 'buildifier'.  does that mean that the BUILD file is not alphabetical or something similar?

**Comment by DavidNorman** [comment]
ah ha - i found the log (was hidden in a 'show whole log' thing) - will fix up

**Comment by kayzhu** [comment]
@tensorflow-jenkins test this please

**Comment by kayzhu** [comment]
As far as I can tell, there's only compile error to fix:

https://ci.tensorflow.org/job/tensorflow-pull-requests-xla/1130/consoleFull

tensorflow/compiler/plugin/executor/executable.cc: In function 'perftools::gputools::DeviceMemoryBase xla::executorplugin::AllocateOutputBuffer(perftools::gputools::executorplugin::ExecutorExecutor*, xla::Literal*)':
tensorflow/compiler/plugin/executor/executable.cc:54:75: error: no matching function for call to 'xla::Literal::mutable_tuple_literals(tensorflow::int64&)'
           AllocateSingleOutput(executor, literal->mutable_tuple_literals(n));
                                                                           ^

**Review by kayzhu — APPROVED** [review]
Approve, and will merge once the remaining tests are fixed.

**Comment by DavidNorman** [comment]
thanks.   not sure why this is failing

**Comment by DavidNorman** [comment]
but will get to the bottom of it

**Comment by DavidNorman** [comment]
oh i see.   will update to use the new Literal class

**Comment by DavidNorman** [comment]
Hi.  Can we go ahead with this now?  I suspect that the 'needs review' thing might be stopping some merging process from going ahead.

Cheers

:)

**Comment by kayzhu** [comment]
@tensorflow-jenkins test this please

**Comment by DavidNorman** [comment]
I don't believe that any of the failures generated by the last set of checks can be down to this code.  I have not changed the python API, I have not touched sparse tensors, and the other ones seem to be CI faults.

**Comment by kayzhu** [comment]
Yea, they might be have been intermittent.

@hawkinsp can we merge this change?

**Comment by kayzhu** [comment]
@tensorflow-jenkins test this please

**Comment by DavidNorman** [comment]
you guys are up late

**Comment by DavidNorman** [comment]
ah - maybe the problem is that I have not integrated enough of the master to include some changes to the API

**Comment by DavidNorman** [comment]
Hi.

This commit:

```
commit e6f58186363279496c46563e6f065ce7ea16c501
Author: Bo Wang <david.b.wang@gmail.com>
Date:   Mon Jun 5 11:41:32 2017 -0700
```

contains addition to the public python API without updating the associated golden API list.  I don't know how it got through the smoke tests, but somehow it did.

**Comment by DavidNorman** [comment]
Not sure about the Windows CMAKE tests.  The failing ones are standard core python sparse ops tests.  I'm certain that your evaluator isn't being used as the device for these tests, because there is no XLA enabled. 

strange

**Comment by DavidNorman** [comment]
the 'ci.tensorflow.org' is just a collection of the other ones - including the 2 failures.

**Comment by DavidNorman** [comment]
The Windows CMAKE test is failing for several other public pull requests (picked 3 at random).

I have failed an Issue.

**Comment by DavidNorman** [comment]
I would say, looking at the history of the Windows CMAKE tests that it fails randomly, and has been doing so for quite a while.

**Comment by kayzhu** [comment]
Yea, I'm currently looking at the failure in Linux CPU Tests:

```
ERROR:tensorflow:TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

ERROR:tensorflow:2 differences found between API and golden.
ERROR:tensorflow:Issue 1	: Change detected in python object: tensorflow.
ERROR:tensorflow:Issue 2	: New object tensorflow.LMDBReader found (added).
F.
======================================================================
FAIL: testAPIBackwardsCompatibility (__main__.ApiCompatibilityTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/tensorflow-pull-requests-cpu/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace/bazel-out/local-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 229, in testAPIBackwardsCompatibility
    update_goldens=FLAGS.update_goldens)
  File "/var/lib/jenkins/workspace/tensorflow-pull-requests-cpu/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace/bazel-out/local-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 188, in _AssertProtoDictEquals
    self.fail('%d differences found between API and golden.' % diff_count)
AssertionError: 2 differences found between API and golden.
```

I do notice tensorflow/tools/api/tests:api_compatibility_test passing from the tip of the tree. I wonder if you could integrate the tip and re-run the tests?

**Comment by kayzhu** [comment]
@tensorflow-jenkins test this please

**Comment by kayzhu** [comment]
Regarding the ABI compatibility failure: I was told it has been fixed this morning, so syncing to the tip and re-run should make that failure go away.

**Comment by kayzhu** [comment]
@martinwicke the existing failures Linux CPU Tests (Python 3) and Windows Cmake Tests (https://ci.tensorflow.org/job/tensorflow-pull-requests-multijob/5415/) seems to be flaky.

Is it safe to merge? And if so, could you merge? Thanks!

**Pull request merged**
